Package Help dialog updated

CSMathematics opened the pull request:
Updated CTAN package database with more information for each package.
![TexDialog](https://github.com/texstudio-org/texstudio/assets/53839524/7bd08b41-599d-477a-a988-44623e495e03)

sunderme commented:
looks good

* packageList.xml can be removed from qrc and git if we take this solution.
* where is the json database coming from ?
* the dialog is a bit small on my 4k monitor, maybe that can be optimized (especially as it does not remember resized dialogs)

sunderme commented on src/texdocdialog.h:
```diff
@@ -38,6 +39,7 @@ private slots:
 	QString lastDocRequest;
     Help *help;
     QStringList m_packages;
+    QButtonGroup ButtonGroup;
```
to follow (implicit) naming strategy, the variable should start with a lower case letter (buttonGroup or even m_buttonGroup)

CSMathematics replied:
Good point. I will review the code again to make these changes.

sunderme commented on src/texdocdialog.ui:
```diff
@@ -6,32 +6,72 @@
+         <number>0</number>
+        </property>
+        <item>
+         <layout class="QHBoxLayout" name="LanguagesLayout">
```
the ususal naming strategy is something like lyLanguages or similar.
lower case first letter helps readability.

sunderme commented on src/texdocdialog.ui:
```diff
@@ -6,32 +6,72 @@
+         </layout>
+        </item>
+        <item>
+         <widget class="QTextBrowser" name="PackageDescriptions"/>
```
see comment above

CSMathematics commented:
The json database is drawn from [CTAN](https://www.ctan.org/help/json/2.0/packages) but it has been edittted.
I merged 4 different json files into 1 database file by building a custom function that gets the data and creating the file. It is not included in this project. Ithink I will be able to upload it in my repository in a few days.

CSMathematics commented:
The only thing I couldn't solve is to find a way to open the package documentation when the link is clicked inside the QTextBrowser.

CSMathematics commented:
All changes are made.

sunderme commented:
does not compile on OSX (LLVM)

sunderme commented on src/texdocdialog.cpp:
```diff
@@ -140,7 +143,28 @@ void TexdocDialog::itemChanged(QTableWidgetItem* item)
+        delete bt;
+    }
+    for(CTANDescription description : package.descriptions){
+        QPushButton * LangButton = new QPushButton(description.language,this);
```
*L*angbutton

sunderme commented on src/texdocdialog.cpp:
```diff
@@ -140,7 +143,28 @@ void TexdocDialog::itemChanged(QTableWidgetItem* item)
+
+    while(buttonGroup.buttons().count()>0){
+        QAbstractButton * bt = buttonGroup.buttons().first();
+        ui->LanguagesLayout->removeWidget(bt);
```
LanguageLayout should start with lower case

CSMathematics replied:
Missed that. I will fix it right away

CSMathematics commented:
> does not compile on OSX (LLVM)

CSMathematics commented:
> > does not compile on OSX (LLVM)

I changed the variable names. But I can't see the reason that compilation fails on OSX. I looked at the compile log and if I'm not mistaken the errors are on other files than the ones I modified.

sunderme commented:
thanks

tmelorc commented:
There are blank spaces before the colon. 
For example, `Name : a0poster`

octaeder commented:
As discussed with latex help and txs manual, it is preferred to hold sources of information locally. So the question is what happens here when you have no internet connection? Can txs handle this? Is the list of packages including captions still available? Has anyone tested this case?

tmelorc commented:
In addition, I tried to compile it now:
```
-- The current git hash is a8c0199
-- The current git describe is 4.4.0-1053-ga8c01995f
```

But I got the error:
![Seleção_101](https://github.com/texstudio-org/texstudio/assets/15181573/aa0312c1-12ad-4031-8c6d-25a41770f8eb)

octaeder commented:
I was able to build with msys2 ucrt64

octaeder commented:
TeXstudio 4.7.3 (git 4.7.2-38-ga8c01995f)
Using Qt Version 6.6.1, compiled with Qt 6.6.1 R

tmelorc commented:
> TeXstudio 4.7.3 (git 4.7.2-38-ga8c01995f) Using Qt Version 6.6.1, compiled with Qt 6.6.1 R

It seems that `setMarkdown` is available only for Qt6.6.

octaeder commented:
Is this the same?
https://doc.qt.io/qt-6.2/qtextdocument.html#setMarkdown
https://doc.qt.io/qt-5/qtextdocument.html#setMarkdown

tmelorc commented:
> Is this the same? https://doc.qt.io/qt-6.2/qtextdocument.html#setMarkdown https://doc.qt.io/qt-5/qtextdocument.html#setMarkdown

Looks like. But the doc says _since 5.14_. Let me check my Qt version. Maybe I have older one.

dbitouze commented:
Nice feature!

But I don't understand the point of by default disabling the search text field: as long as the “all the packages” option is disabled (the corresponding checkbox is unchecked), nothing can be done. What I would expect would be that, by default:

- the “all the packages” option is disabled
- the search text field is enabled, allowing the user to search for a particular package (without cluttering the interface with the list of all packages).

octaeder commented:
@dbitouze this is impossible for you?

![image](https://github.com/texstudio-org/texstudio/assets/102688820/fb02932f-088c-47e4-b439-21053411af7d)

and I would say that the option is always disabled when the dialog is started

tmelorc commented:
I'm using the version below and the text field is locked, as screenshot below.
```
TeXstudio 4.7.3 (git 4.7.1-81-gf0f4089a9)
Utilizando Qt Versão 5.15.3, compilado com Qt 5.15.3 R

```
![Documentação dos Pacotes (Texdoc)_002](https://github.com/texstudio-org/texstudio/assets/15181573/cd823413-cfb7-4d09-9228-09d7332af169)

octaeder commented:
Including 5 extra developments and missing the latest 2 (reagarding cwls) I use this (msys2 build) on Win10:

TeXstudio 4.7.3 (git 4.7.2-51-gbd8a0bd91)
Using Qt Version 6.6.1, compiled with Qt 6.6.1 R

I wonder why you see no packages at all. In any case the packages list should not be empty at start. How do you get the packages into place?

tmelorc commented:
> I wonder why you see no packages at all. In any case the packages list should not be empty at start. How do you get the packages into place?

Sorry, I didn't follow the question.

tmelorc commented:
The packages list only appears after the check button is on. Then the text field is unlocked.

![Documentação dos Pacotes (Texdoc)_003](https://github.com/texstudio-org/texstudio/assets/15181573/bae95571-9fcc-4a7d-b2c3-a628a41add2d)

octaeder commented:
there's something wrong.

tmelorc commented:
> there's something wrong.

Maybe this line: `ui->lineEditSearch->setEnabled(n>0);`

octaeder commented:
since search field is not enabled I would conclude that n equals 0. But then m_packages is empty & state<=0. Since I can't reproduce, can can you debug the the values of m_packages.count() and state immediately after line 48 (int n=0;)?

octaeder commented:
btw: do you use packages in your latex document?

tmelorc commented:
> btw: do you use packages in your latex document?

You got it! I was opening the texdoc window with no opened document.
Now it works.

Anyway, it would be good to have the feature available also for no documents.

octaeder commented:
in this case the option should be checked
